### PR TITLE
Add support for non-responsive values to `Grid`'s `gap`, `columns`, and `areas` props.

### LINKED ISSUE #10484: Align Grid with other layout components
Grid was built before the layout foundations project and was scheduled for deprecation before realizing other Shopify systems will include a grid component. We need to bring it up to other layout component standards like having short hand properties for columns, and gap, not just responsive values

### PR DESCRIPTION
**Pull request by jesstelford**
Builds on functionality introduced in https://github.com/Shopify/polaris/pull/10404.

In service of bringing `Grid`'s responsive props in line with the rest of the system, I also standardized some other parts of the system along the way.

I purposely kept these as individual commits for easier reviewing and also so they retain their history, meaning **we should do fast-forward merge instead of a merge commit** ("Rebase and merge" in GitHub's UI) so the commits don't get squashed.

See the `commits` tab for the changes.

### CODE REVIEW

File: .changeset/slimy-donuts-report.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@shopify/polaris': minor
+---
+
+Add support for non-responsive values to `Grid`'s `gap`, `columns`, and `areas` props.
```
Comment (kyledurand) [inline .changeset/slimy-donuts-report.md]:
```suggestion
Added support for non-responsive values to `Grid`'s `gap`, `columns`, and `areas` props.
```

File: polaris-react/src/utilities/css.ts [hunk 2]
```diff
@@ -52,57 +86,101 @@ export function sanitizeCustomProperties(
...
+  };
+}
+
+export function mapResponsivePropValues<Input, Output>(
+  responsiveProp: ResponsiveProp<Input> | undefined,
+  fn: (value?: Input) => Output | undefined,
+): ResponsiveProp<Output> | undefined {
+  if (isObject(responsiveProp)) {
+    return Object.fromEntries(
+      (Object.entries(responsiveProp) as Entries<typeof responsiveProp>).map(
```
Comment (kyledurand) [inline polaris-react/src/utilities/css.ts:179]: Since we're only using Entries from type-fest in this one spot, and we know the shape, and we're casting, I think we can do away with the type-fest import and type this manually.

File: polaris-react/src/components/InlineGrid/InlineGrid.tsx [hunk 3]
```diff
@@ -23,18 +21,18 @@ export interface InlineGridProps extends React.AriaAttributes {
    * columns={6}
    * columns={{xs: 1, sm: 1, md: 3, lg: 6, xl: 6}}
    */
-  columns?: Columns;
+  columns?: ResponsiveProp<ColumnsType>;
```
Comment (kyledurand) [inline polaris-react/src/components/InlineGrid/InlineGrid.tsx:24]: This might affect props generation but not 100% sure. Can you compare the docs site with a local build to see what the props table shows?

**Review by kyledurand — COMMENTED** [review]
It's hard to tell what changes to some of our utilities are necessary here and I'm afraid the scss changes only confuse things more, and also make it harder to migrate away from scss or to a post css processor. 

Can you outline why these changes were made and why you couldn't fix the bug(s) by using existing utilities? I'm talking specifically about the changes in  `_responsive-props.scss`, `css.ts`

**Comment by kyledurand** [comment]
Just noticed you want to merge this into the next branch. Shouldn't bug fixes / minor changes go into main?

We'll also want to `/snapit` and compare admin areas where [Grid is used](https://polaris-coverage.shopifycloud.com/usage/component?repo=shopify%2Fweb&type=components&element=%3CGrid%3E)

**Comment by kyledurand** [comment]
We need to ship this or a simplified version

**Comment by jesstelford** [comment]
> Shouldn't bug fixes / minor changes go into main?

Yeah, but in this case I wanted to align with the changes in `next` without introducing a whole bunch of merge conflicts, and figured since we're rolling out v12 real soon anyway, it makes it easier to skip `main`.

File: polaris-react/src/utilities/css.ts (hunk 2 above)
Comment (jesstelford) [reply polaris-react/src/utilities/css.ts:179]: I really don't want to maintain the `Entries` type - it's complicated enough that I barely understand it, but I do know it does exactly what I needed, and I landed on this solution after attempting to manually type it first (though admittedly I quickly rage quit because I'd had enough of TypeScript's cryptic errors and nonesense inability to understand basic things like JS's `Object.entries`).

**Comment by jesstelford** [comment]
> It's hard to tell what changes to some of our utilities are necessary here

Fair enough, there's a lot going on in this PR. The commits were split up to try make it clear which changes are related to altering the utilities and for what purpose. But to expand on each of them:

1. `Support default CSS variables beyond 'initial'`
    - We're leveraging CSS Vars property of being DOM-inherited instead of Cascade-inherited (ie; it will crawl DOM nodes looking for CSS custom properties, instead of crawling the CSS cascade)
    - The CSS `var()` lookup will stop whenever _any_ value is found. `initial` is a valid value, so the lookup algorithm stops. 
    - But when `initial` is used in a `var`, it will cause that property to not be applied to the final calculated CSS.
    - With that in place, we've got a (basic) emulation of CSS scope while it's still being spec'd.
    - There are cases where we want the component to have a valid "default" value for some properties. There are two ways to do this:
         1. `--some-var: initial; --some-var: <default>;` Where we always set `initial` to enforce the "scope", then manually set the "<default>" value after, or;
         2. `--some-var: <default>;` Where we just set the value to the default at the start.
    - At first it doesn't seem like much of a difference until you consider setting different defaults for each breakpoint. Option 1 is then duplicating the custom property for every breakpoint which is wasteful and confusing when inspecting the DOM. However, Option 2 is much clearer and has less duplication.
    - Since we're already configuring responsive props with a single mixin call, it makes sense to also configure the default values at the same time.
    - Later, this is used by `Grid` which has different defaults per breakpoint in a simple `$default: ('xs': 6, 'lg': 12)`

2. Standardize how we map over responsive props within components
    - This is just removing duplicate code and cleaning up the function signature in the process

3. Tighten up the types and deduplicate CSS Var generation code
    - I wish this commit didn't have to exist, but since we're using TS, there were a handful of cases where it was either not enforcing types tightly enough, or was giving erroneous errors when used by Grid.
    - Basically it's a TS bugfix commit.

4. Add support for non-responsive values to Grid's gap, columns, and `areas` props
    - This was the ultimate goal of these changes, and fixes the linked issue.
    - It greatly simplifies the SASS of Grid, standardises it along side all our other components, and leverages the previous 3 commits to do so.

> the scss changes only confuse things more

Interesting! I find the new format of the SCSS file to be much easier to read and understand with much reduced duplication and chance for human error to sneak in. It also standardizes it with all our other components, so ideally once you've groked one, you've groked them all.

> also make it harder to migrate away from scss or to a post css processor.

I'd argue the abstraction makes it _easier_; Instead of having to manually migrate each instance where we have multiple breakpoints with `initial`s being set, etc, we've now got a single mixin which needs to be changed to whatever a new system might need. For example, switching to Vanilla Extract would mean simply rewriting the `responsive-props` mixin in JS, and shift the callsites to the .tsx files instead of .scss files instead of having to copy/paste huge chunks of arbitrary media queries, etc.

File: polaris-react/src/components/InlineGrid/InlineGrid.tsx (hunk 3 above)
Comment (jesstelford) [reply polaris-react/src/components/InlineGrid/InlineGrid.tsx:24]: Comes out identically 👍 

Screenshot 2023-10-04 at 6 10 01 pm

**Comment by jesstelford** [comment]
> We'll also want to `/snapit` and compare admin areas where [Grid is used](https://polaris-coverage.shopifycloud.com/usage/component?repo=shopify%2Fweb&type=components&element=%3CGrid%3E)

I'll definitely need help with this one - I often struggle to get spin to do what I need / find those places of usage actually in the Admin.

File: polaris-react/src/components/Grid/Grid.scss [hunk 4]
```diff
@@ -1,60 +1,24 @@
...
+    'gap',
+    'gap',
+    $default: 'var(--p-space-400)'
+  );
+  @include responsive-props('grid', 'areas', 'grid-template-areas');
+  @include responsive-props(
+    'grid',
+    'columns',
+    '--pc-grid-template-columns',
+    $default: ('xs': 6, 'lg': 12)
```
Comment (kyledurand) [inline polaris-react/src/components/Grid/Grid.scss:17]: Can we override these defaults when only a single value gets passed in?

**Review by kyledurand — CHANGES_REQUESTED** [review]
I think this is really close. My main concern now is that when passing in a non-responsive value it doesn't override the responsive defaults which will lead to confusion.

image

Secondary concern is that adding to the mixin complexity will add tech debt when trying to remove sass. Do you have a plan for that? I think we might be removing sass sooner rather  than later and not sure post css conversion will work 1:1. This is not a blocker

File: polaris-react/src/components/Grid/Grid.scss (hunk 4 above)
Comment (jesstelford) [reply polaris-react/src/components/Grid/Grid.scss:17]: Ooh, good catch - I missed that case. I'll patch that up this morning 👍 

> add tech debt when trying to remove sass.

Depends on what we're aiming for;

1. If we're going to PostCSS, then most of the functionality can be [moved into `postcss-mixins`](https://github.com/postcss/postcss-mixins) by the looks of it.
2. If we're going to CSS-in-JS, then the functionality can be converted to JS fairly easily.
3. If we're going to pure CSS without PostCSS plugins or CSS-in-JS then we could either fallback to creating and injecting stylesheets, or reassess the overall strategy.

Comment (jesstelford) [reply polaris-react/src/components/Grid/Grid.scss:17]: Was able to fix the defaults issue with @gwyneplaine's help 🎉 

I amended an earlier commit, so I've put the diff below for you to review.

tl;dr: Reverted back to using `initial` for scoping the CSS custom properties, then moved the default into the media queries (where they should have been all along). The new `find-default-with-fallback()` function matches the semantics of the consumer-facing responsive props API. ie; things like `$default: ('sm': 4, 'lg': 8)` are effectively the same as: `$default: ('xs': null, 'sm': 4, 'md': 4, 'lg': 8, 'xl': 8)`.

```diff
diff --git a/polaris-react/src/styles/shared/_responsive-props.scss b/polaris-react/src/styles/shared/_responsive-props.scss
index 319a70f2a..1ae4fc569 100644
--- a/polaris-react/src/styles/shared/_responsive-props.scss
+++ b/polaris-react/src/styles/shared/_responsive-props.scss
@@ -6,11 +6,37 @@
   @return type-of($var) == 'null';
 }
 
+// NOTE: Order is important here; will search backwards through this list to
+// find fallback values
+$breakpointKeys: ('xs', 'sm', 'md', 'lg', 'xl');
+
+@function find-default-with-fallback($default, $breakpoint) {
+  $initialIndex: index($breakpointKeys, $breakpoint);
+
+  // If the given '$breakpoint' doesn't exist, just return null
+  @if is-null($initialIndex) {
+    @return null;
+  }
+
+  // Search backwards until a non-null value is found in the $defaults map
+  // NOTE: Arrays in SASS are 1-based, so we're iterating down to AND INCLUDING
+  // index "1" (the first element).
+  @for $index from $initialIndex through 1 {
+    $value: map-get($default, nth($breakpointKeys, $index));
+    @if not is-null($value) {
+      @return $value;
+    }
+  }
+
+  // Couldn't find a non-null default, so return null
+  @return null;
+}
+
 @function get-default($default: null, $key: null) {
   $value: null;
   @if is-map($default) {
     @if not is-null($key) {
-      $value: map-get($default, $key);
+      $value: find-default-with-fallback($default, $key);
     }
   } @else {
     $value: $default;
@@ -21,18 +47,17 @@
 @mixin scope-custom-property(
   $componentName,
   $componentProp,
-  $responsively: false,
-  $default: null
+  $responsively: false
 ) {
   @if $responsively {
     // stylelint-disable -- Polaris component custom properties
-    --pc-#{$componentName}-#{$componentProp}-xs: #{get-default($default, 'xs')};
-    --pc-#{$componentName}-#{$componentProp}-sm: #{get-default($default, 'sm')};
-    --pc-#{$componentName}-#{$componentProp}-md: #{get-default($default, 'md')};
-    --pc-#{$componentName}-#{$componentProp}-lg: #{get-default($default, 'lg')};
-    --pc-#{$componentName}-#{$componentProp}-xl: #{get-default($default, 'xl')};
+    --pc-#{$componentName}-#{$componentProp}-xs: initial;
+    --pc-#{$componentName}-#{$componentProp}-sm: initial;
+    --pc-#{$componentName}-#{$componentProp}-md: initial;
+    --pc-#{$componentName}-#{$componentProp}-lg: initial;
+    --pc-#{$componentName}-#{$componentProp}-xl: initial;
   } @else {
-    --pc-#{$componentName}-#{$componentProp}: #{get-default($default)};
+    --pc-#{$componentName}-#{$componentProp}: initial;
   }
 }
 
@@ -42,14 +67,12 @@
   $declarationProp,
   $default: null
 ) {
-  @include scope-custom-property(
-    $componentName,
-    $componentProp,
-    false,
-    $default
-  );
+  @include scope-custom-property($componentName, $componentProp, false);
 
-  #{$declarationProp}: var(--pc-#{$componentName}-#{$componentProp});
+  #{$declarationProp}: var(
+    --pc-#{$componentName}-#{$componentProp},
+    #{get-default($default)}
+  );
 }
 
 @mixin responsive-props(
@@ -58,19 +81,20 @@
   $declarationProp,
   $default: null
 ) {
-  @include scope-custom-property(
-    $componentName,
-    $componentProp,
-    true,
-    $default
-  );
+  @include scope-custom-property($componentName, $componentProp, true);
 
-  #{$declarationProp}: var(--pc-#{$componentName}-#{$componentProp}-xs);
+  #{$declarationProp}: var(
+    --pc-#{$componentName}-#{$componentProp}-xs,
+    #{get-default($default, 'xs')}
+  );
 
   @media #{$p-breakpoints-sm-up} {
     #{$declarationProp}: var(
       --pc-#{$componentName}-#{$componentProp}-sm,
-      var(--pc-#{$componentName}-#{$componentProp}-xs)
+      var(
+        --pc-#{$componentName}-#{$componentProp}-xs,
+        #{get-default($default, 'sm')}
+      )
     );
   }
 
@@ -79,7 +103,10 @@
       --pc-#{$componentName}-#{$componentProp}-md,
       var(
         --pc-#{$componentName}-#{$componentProp}-sm,
-        var(--pc-#{$componentName}-#{$componentProp}-xs)
+        var(
+          --pc-#{$componentName}-#{$componentProp}-xs,
+          #{get-default($default, 'md')}
+        )
       )
     );
   }
@@ -91,7 +118,10 @@
         --pc-#{$componentName}-#{$componentProp}-md,
         var(
           --pc-#{$componentName}-#{$componentProp}-sm,
-          var(--pc-#{$componentName}-#{$componentProp}-xs)
+          var(
+            --pc-#{$componentName}-#{$componentProp}-xs,
+            #{get-default($default, 'lg')}
+          )
         )
       )
     );
@@ -106,7 +136,10 @@
           --pc-#{$componentName}-#{$componentProp}-md,
           var(
             --pc-#{$componentName}-#{$componentProp}-sm,
-            var(--pc-#{$componentName}-#{$componentProp}-xs)
+            var(
+              --pc-#{$componentName}-#{$componentProp}-xs,
+              #{get-default($default, 'xl')}
+            )
           )
         )
       )
```

Comment (gwyneplaine) [reply polaris-react/src/components/Grid/Grid.scss:17]: @kyledurand here's a [local reproduction url](http://localhost:3000/sandbox?code=N4Igxg9gJgpiBc4IDsDOAXABAcQE4EsoBhAC3wBspcZlMBeTACgEp6A%2BJgHWQB43vMmHnkIA6IjHLl%2BtQUKIBDXFDZEI5TAEYeAekXKZg3SKjjJ0gUJNmphuT30q1GgEy7Hd4wVMTbl4d42FrJGHs6YAMzuSir%2BOta%2BwUYJ5nahMarqmAAs0QZxKX6yAWKJafIZ4QCsebHF8YFl-oVJ9mFZAGy1ng2lqdy6MszcAwAKCgDmMJ4AQuQQYADWAMroCkuYEwoADnScINkADIf75TwAKjAAHuhsMwqoMJhgDzC6lzdnJuXJ3qQUVBoOjOvTqciEH1uADkUJhqKhtihUPgAG5vHSQr7eZ7qACuAFs0HsQC5TpZwSViGRKNRaMDyUJQWdMQAlGAIpGop6KR6od7XW4Myk48gEonAYCYK6oeCYfakkCYAC%2BSp%2BVj%2B1MBdJB3yFJhFYtQdAlUplcpJ%2BwANJhyBNZftsvtlaqGb9CP8aTRMPSQozdb6LgK2GyOWguZh8MgSDACGtkGB0ZjXcLIKLCUaTfioPaQBEnSq1er3ZraT7wX7vD05gsVmsljJdOMpjIQJaQOho-j2QgANogeYvcgweA0EAAXTbAHdCB3UL2x0qgA) with the above cases in playroom.

**Review by gwyneplaine — COMMENTED** [review]
(unofficially) lgtm ✅

**Comment by jesstelford** [comment]
@kyledurand of note in this PR is a bug fix for the differing behaviour of `columns`, `areas`, and `gap` props on Grid to make their responsiveness align correctly with the semantics of our responsive props across Polaris. It's to do with the way we fallback to lower and lower breakpoint values if the current breakpoint doesn't have an explicit value.

## `columns` and `areas`

Current behaviour:

1. If set, use an explicit value passed for the current breakpoint. Else, go to Step 2.
2. If set, use a default value for the current breakpoint. Else, go to Step 3.
3. If not at at `xs` breakpoint, go down one breakpoint and go to Step 1. Else, go to Step 4.
4. Set the value to `initial`, resulting in not applying the property.

New behaviour:

1. If set, use an explicit value passed for the current breakpoint. Else, go to Step 2.
2. If not at at `xs` breakpoint, go down one breakpoint and go to Step 1. Else, go to Step 3.
3. If set, use a default value for the current breakpoint. Else, go to Step 4.
4. Set the value to `initial`, resulting in not applying the property.

Another way to think about it is how the props are expanded to have concrete values for each breakpoint:

For example, given a default of `('xs': 6, 'lg': 12)`, the prop values expand as such:

| `columns=`           | Current behaviour                              | New behaviour                   |
|:----------------------|--------------------------------------------------------|-------------------------------------------|
| `3`                  | n/a                                                    | `{xs: 3, sm: 3, md: 3, lg: 3, xl: 3}` |
| `{md: 3}`        | `{xs: 6, sm: 6, md: 3, lg: 12, xl: 12}` | `{xs: 6, sm: 6, md: 3, lg: 3, xl: 3}` |
| `{md: 3, lg: 2}` | `{xs: 6, sm: 6, md: 3, lg: 2, xl: 2}`       | `{xs: 6, sm: 6, md: 3, lg: 2, xl: 2}` |
| `{md: 3, xl: 1}` | `{xs: 6, sm: 6, md: 3, lg: 12, xl: 1}`       | `{xs: 6, sm: 6, md: 3, lg: 3, xl: 1}` |

## `gap`

Before this PR, `gap` was implemented differently to `columns` and `areas`:

Current behaviour:

1. If set, use an explicit value passed for the current breakpoint. Else, go to Step 2.
4. Set the value to the default.

New behaviour: _(same as for `columns` and `areas`)_

1. If set, use an explicit value passed for the current breakpoint. Else, go to Step 2.
2. If not at at `xs` breakpoint, go down one breakpoint and go to Step 1. Else, go to Step 3.
3. If set, use a default value for the current breakpoint. Else, go to Step 4.
4. Set the value to `initial`, resulting in not applying the property.

For example, given the default of `4`, the prop values expand as such:

| `gap=`           | Current behaviour                              | New behaviour                   |
|:----------------------|--------------------------------------------------------|-------------------------------------------|
| `3`                  | n/a                                                    | `{xs: 3, sm: 3, md: 3, lg: 3, xl: 3}` |
| `{md: 3}`        | `{xs: 4, sm: 4, md: 3, lg: 4, xl: 4}` | `{xs: 4, sm: 4, md: 3, lg: 3, xl: 3}` |
| `{md: 3, lg: 2}` | `{xs: 4, sm: 4, md: 3, lg: 2, xl: 4}`       | `{xs: 4, sm: 4, md: 3, lg: 2, xl: 2}` |
| `{md: 3, xl: 1}` | `{xs: 4, sm: 4, md: 3, lg: 4, xl: 1}`       | `{xs: 4, sm: 4, md: 3, lg: 3, xl: 1}` |

**Comment by kyledurand** [comment]
Thanks for the breakdown @jesstelford I'm pretty good with the changes, but we need to snap this and make sure it at least passes web ci before merging. Just trying to sort that out now

**Comment by jesstelford** [comment]
Looks like [the snapshots are failing to publish](https://github.com/Shopify/polaris/actions/runs/6431771117/job/17465302921), but I'm not sure why... Is it because the base is `next`? How're we doing snapshots on other branches into `next`?

**Comment by jesstelford** [comment]
@laurkim [noticed an issue with `color` being defaulted incorrectly](https://github.com/Shopify/polaris/pull/10890#pullrequestreview-1661701000):

> Noticing some regressions with text when running locally though where the color isn't being set properly.

I've tracked it down to the difference between `initial` and `inherit` for `color`. This would have been an existing bug in our system, except for a compounding bug which cancelled it out:

When a `color` prop isn't passed, the CSS custom property's value is set to `undefined`, which means it's not set on the `style` attribute, so the declaration of `color: var(--pc-box-color)` becomes invalid, and falls back to the user agent's default of `color: inherit`.

The new behaviour in this PR enforces consistency in CSS custom properties, and set the default to `initial` always. While this was technically the most correct approach for scoping our variables (ie; if you want nested text to have the same `color`, you'd have to specify it explicitly), it doesn't match our consumer's expectation of how CSS works: It is expected that some properties will inherit from their parent automatically (eg; `color`, `font`, etc). So, I've fixed that by setting the default to `inherit` for the properties which need it based on data from https://www.npmjs.com/package/@webref/css.

[Here's the diff of changes](https://github.com/Shopify/polaris/compare/aa81906a1897f46a21e6726bcacd3227972827a5..89610fb982eef822c0e1a582f840b8077ca53d00) made to ensure some declarations are inherited.

---

An alternative implementation might be to mirror the existing behaviour of not passing the custom property through to the `style` attribute when a value isn't set.

If we went down this path, we'd need to move the "default" values up into the JS, then pass those down to the media queries via `style` tags. Unfortunately that would mean we'd be passing those defaults down for every _instance_ of a component instead of just once per _definition_ of a component. For example, `Box` has a default `border-width: 0`. If the page renders 45 `<Box>`es, there would potentially be 45 instances of `style="--pc-box-border-width-default: 0"` as opposed to just setting `--pc-box-border-width: 0` once in the generated `Box.css` file.

**Comment by jesstelford** [comment]
🤔 I can't figure out what's causing these "19 changes" in Chromatic...

Screenshot 2023-10-08 at 9 06 21 am

The screenshots seem to indicate that vertical padding is different, but when I load up the snapshots ([base](https://shopify.chromatic.com/test?appId=5d559397bae39100201eedc1&id=6521d4926312cb3608d603d0&k=652050e05cd603ad14f602d9-1200px-interactive-true) vs [this PR](https://shopify.chromatic.com/test?appId=5d559397bae39100201eedc1&id=6521d4926312cb3608d603d0&k=6521d4926312cb3608d603d0-1200px-interactive-true)), there's no difference, and I can't replicate it locally.

I'm wondering if it's a timing issue of some kind with client side rendering 🤔 🤷
